File: 50-strukture/binarno-stablo.js
class Cvor {
  constructor(vrednost) {
    this.vrednost = vrednost
    this.levo = null
    this.desno = null
  }
}

class Stablo {
  constructor(vrednost) {
    this.koren = new Cvor(vrednost)
  }

  dodaj(vrednost, cvor = this.koren) {
    if (vrednost === cvor.vrednost) return
    const strana = vrednost > cvor.vrednost ? 'desno' : 'levo'
    if (!cvor[strana]) cvor[strana] = new Cvor(vrednost)
    else this.dodaj(vrednost, cvor[strana])
  }

  ispisi(cvor = this.koren) {
    if (!cvor) return
    this.ispisi(cvor.levo)
    console.log(cvor.vrednost)
    this.ispisi(cvor.desno)
  }

  ispisiNeredjano(cvor = this.koren) {
    if (!cvor) return
    console.log(cvor.vrednost)
    this.ispisi(cvor.levo)
    this.ispisi(cvor.desno)
  }

  ispisiPoSirini(cvor = this.koren) {
    const red = cvor ? [cvor] : []
    while (red.length) {
      const cvor = red.shift()
      console.log(cvor.vrednost)
      if (cvor.levo) red.push(cvor.levo)
      if (cvor.desno) red.push(cvor.desno)
    }
  }

  ispisiPoDubini(cvor = this.koren) {
    const stog = cvor ? [cvor] : []
    while (stog.length) {
      const cvor = stog.pop()
      console.log(cvor.vrednost)
      if (cvor.levo) stog.push(cvor.levo)
      if (cvor.desno) stog.push(cvor.desno)
    }
  }

  ispisiPoDubiniRekurzivno(cvor = this.koren) {
    if (!cvor) return
    console.log(cvor.vrednost)
    this.ispisi(cvor.desno)
    this.ispisi(cvor.levo)
  }
}

const stablo = new Stablo(5)
stablo.dodaj(3)
stablo.dodaj(7)
stablo.dodaj(4)
stablo.dodaj(2)
stablo.dodaj(3) // odbija isti

console.log(JSON.stringify(stablo, null, 4))
console.log('ispisuje redom:')
stablo.ispisi()
console.log('ispisuje neredjano:')
stablo.ispisiNeredjano()
console.log('ispisuje po sirini:')
stablo.ispisiPoSirini()
console.log('ispisuje po dubini:')
stablo.ispisiPoDubini()

File: 50-strukture/binarno-stablo/binarno-stablo.py
class Cvor:
    def __init__(self, val):
        self.levo = None
        self.desno = None
        self.data = val


def dodaj(koren, cvor):
    if koren is None:
        koren = cvor
    else:
        if koren.data > cvor.data:
            if koren.levo is None:
                koren.levo = cvor
            else:
                dodaj(koren.levo, cvor)
        else:
            if koren.desno is None:
                koren.desno = cvor
            else:
                dodaj(koren.desno, cvor)


def stampaj_poredjano(koren):
    if not koren:
        return
    stampaj_poredjano(koren.levo)
    print koren.data
    stampaj_poredjano(koren.desno)


def stampaj_neredjano(koren):
    if not koren:
        return
    print koren.data
    stampaj_neredjano(koren.levo)
    stampaj_neredjano(koren.desno)


stablo = Cvor(3)
dodaj(stablo, Cvor(7))
dodaj(stablo, Cvor(1))
dodaj(stablo, Cvor(5))

print "poredjano:"
stampaj_poredjano(stablo)

print "neredjano:"
stampaj_neredjano(stablo)
